docs/src/callbacks.md

```diff
@@ -0,0 +1,168 @@
+```@meta
+CurrentModule = JuMP
+DocTestSetup = quote
+    using JuMP
+end
+DocTestFilters = [r"≤|<=", r"≥|>=", r" == | = ", r" ∈ | in ", r"MathOptInterface|MOI"]
+```
+
+# Callbacks
+
+Many mixed-integer (linear, conic, and nonlinear) programming solvers offer
+the ability to modify the solve process. Examples include changing branching
+decisions in branch-and-bound, adding custom cutting planes, providing custom
+heuristics to find feasible solutions, or implementing on-demand separators to
+add new constraints only when they are violated by the current solution (also
+known as lazy constraints).
+
+While historically this functionality has been limited to solver-specific
+interfaces, JuMP provides solver-independent support for three types of
+callbacks:
+
+ 1. lazy constraints
+ 2. user-cuts
+ 3. heuristic solutions
```

Comment:
Do callbacks require using `direct_mode`?

Reply:
It should work with cache and bridges too

Reply:
This is untested, at least.

Reply:
There are unit tests in MOI but indeed no integration test at the JuMP level.

Reply:
Hmm. I can't get this to work. I think `OptimizerAttributes` aren't being copied? https://github.com/JuliaOpt/MathOptInterface.jl/blob/1755bf88b05499aced42b4ca1c15267045b038ed/src/Utilities/copy.jl#L331-L340

Reply:
Good point, that will be an issue if the CachingOptimizer state is `EMPTY_OPTIMIZER`, we should fix this in MOI v0.9.8. However, if it is `ATTACHED_OPTIMIZER` then the attributes will be passed to the optimizer during `MOI.set`.

Reply:
Question: why don't we call `attach_optimizer` on the creation of `Model`? Because if we try to add a variable and it isn't supported, it will detach and we're back to the current approach. Whereas the current approach causes a big `copy_to` operation on the first `optimize!`, even if incremental building is supported.

Reply:
The big `copy_to` has the potential to be faster than incremental building because you can pass the problem all at once to efficient methods like `GRBloadmodel`.

Reply:
Yes, the reasoning is that the incremental model building can only be slower than copying it all at once. There is a slight inconsistency with
https://github.com/JuliaOpt/MathOptInterface.jl/blob/1755bf88b05499aced42b4ca1c15267045b038ed/src/Utilities/cachingoptimizer.jl#L173-L175
where we attach the solver when we empty the caching optimizer.